packages/@aws-cdk/cdk/lib/core/construct.ts

```diff
@@ -1,23 +1,40 @@
 import cxapi = require('@aws-cdk/cx-api');
+import { makeUniqueId } from '../util/uniqueid';
 export const PATH_SEP = '/';
 
 /**
  * Represents the building block of the construct graph.
  * When a construct is created, it is always added as a child
  */
 export class Construct {
-    private static readonly VALID_NAME_REGEX = /^[A-Za-z][A-Za-z0-9]*$/;
-
     /**
      * Returns the parent of this node or undefined if this is a root node.
      */
     public readonly parent?: Construct;
 
     /**
-     * The name of this construct
+     * @deprecated use `id`
      */
     public readonly name: string;
 
+    /**
+     * The subtree-local id of the construct.
+     * This id is unique within the subtree. To obtain a tree-unique id, use `uniqueId`.
+     */
+    public readonly id: string;
+
+    /**
+     * The full path of this construct in the tree.
+     * Components are separated by '/'.
```

Comment:
In other words:

> So we had this jolly good array of strings that represent a path in the tree. We could have exposed it as a `ReadonlyArray<string>`, but we though - oh the hell, we can `.join('/')` this so in case you're interested in the array, you'll always have to `.split('/')` it. There there, we fixed it.

![](https://media1.giphy.com/media/SphdK3sFxFYZy/200.gif)

Reply:
Yes, I know not every language supports `ReadonlyArray<T>`, but they could code-generate a copy-on-read and be good with it.

Reply:
The reason this is a string is because I think that in most cases, when people need the path, they want the string version of it and not the components. Optimize for the common case.